### PR TITLE
Allow pushdown of dynamic filters through HashAggregation

### PR DESCRIPTION
**Pull request by xiaodouchen**
Currently, in HashAggregation, only when the aggregation is distinct,
the channel of the group key is added to the identity projections. This
results in a situation where, when the group key matches the join key,
the dynamic filters generated by HashProbe cannot be pushed down through
HashAggregation. This pr supports dynamic filters push down through
HashAggregation when group key and join key is the same.

This optimization applies to plans like:

```
- HashProbe(k1 = k2,...)
  - HashAggregation(k1, sum(x))
    - TableScan
```

### CODE REVIEW

**Comment by xiaodouchen** [comment]
@mbasmanova @xiaoxmeng Could you help to have a look at this PR ?

File: velox/exec/Driver.cpp [hunk 1]
```diff
@@ -68,13 +68,16 @@ class CancelGuard {
...
+    Operator* prevOp) {
   for (const auto& projection : projections) {
     if (projection.outputChannel == outputChannel) {
       return projection.inputChannel;
     }
   }
-  return std::nullopt;
+
+  // If grouping key and join key is the same, continue walking upstream.
+  return prevOp->getGroupKeyInputChannel(outputChannel);
```
Comment (Yuhta) [inline velox/exec/Driver.cpp:80]: This projection is no longer identity.  We don't need identity for filter pushdown to pass through, we only need the projection to be a subset of the its input.  Can we add `Operator::filteredProjections()` and use that for filter pushdown instead?  For most operators this should be forward to `identityProjections()` and for `HashAggregation` you can add all keys.

Comment (xiaodouchen) [reply velox/exec/Driver.cpp:80]: That's a good suggestion.

File: velox/exec/Driver.cpp [hunk 2]
```diff
@@ -984,7 +984,7 @@ std::unordered_set<column_index_t> Driver::canPushdownFilters(
         break;
       }
 
-      const auto& identityProjections = prevOp->identityProjections();
+      const auto& identityProjections = prevOp->filteredProjections();
```
Comment (lingbin) [inline velox/exec/Driver.cpp:987]: rename variable too: `identityProjections` --> `filteredProjections`

File: velox/exec/Operator.h [hunk 3]
```diff
@@ -459,6 +459,17 @@ class Operator : public BaseRuntimeStatWriter {
     return identityProjections_;
   }
 
+  /// Returns a list of filtered projections, which are used for filter pushdown
```
Comment (mbasmanova) [inline velox/exec/Operator.h:462]: > filtered projections

I find this term confusing. Can we find a better name or provide a clearer explanation? Specifically, can you clarify the difference between identityProjection and filteredProjection.

The definition of identity projection "columns that are projected as-is possibly after applying a filter" seems to fit group-by keys. These are returned as is for a subset of rows. Any particular reason HashAgg cannot report group by keys as identity projection? If so, let's update the comment to explain why group-by keys do not qualify.

File: velox/exec/Operator.h [hunk 4]
```diff
@@ -459,6 +459,17 @@ class Operator : public BaseRuntimeStatWriter {
...
+  /// to pass through. If the operator has initialized filtered projections,
+  /// such as HashAggregation, return them directly. For most operators, this
+  /// should be forward to identityProjections().
+  const std::vector<IdentityProjection>& filteredProjections() const {
+    if (!filteredProjections_.empty()) {
```
Comment (mbasmanova) [inline velox/exec/Operator.h:467]: Let's remove this field, make this API virtual and have default implementation that returns identityProjections.

**Review by mbasmanova — CHANGES_REQUESTED** [review]
@xiaodouchen This is a nice optimization. Some comments.

File: velox/exec/Operator.h (hunk 3 above)
Comment (xiaodouchen) [reply velox/exec/Operator.h:462]: I don't know the particular reason HashAgg cannot report group by keys as identity projection. I just found that in HashAggregation::initialize(), the identityProjections_ is only initialized for the distinct case, as shown below.

```
  if (isDistinct_) {
    for (auto i = 0; i < hashers.size(); ++i) {
      identityProjections_.emplace_back(hashers[i]->channel(), i);
    }
  }
```

Comment (mbasmanova) [reply velox/exec/Operator.h:462]: > I don't know the particular reason HashAgg cannot report group by keys as identity projection. 

Let's figure out whether this is just an omission or there is a good reason.

File: velox/exec/Operator.h [hunk 5]
```diff
@@ -455,7 +455,7 @@ class Operator : public BaseRuntimeStatWriter {
 
   /// Returns a list of identity projections, e.g. columns that are projected
   /// as-is possibly after applying a filter.
-  const std::vector<IdentityProjection>& identityProjections() const {
+  virtual const std::vector<IdentityProjection>& identityProjections() const {
```
Comment (mbasmanova) [inline velox/exec/Operator.h:458]: Let's not make this change.

If we do, then this should be a separate PR and identityProjections_ needs to become private.

File: velox/exec/HashAggregation.h [hunk 6]
```diff
@@ -50,6 +50,10 @@ class HashAggregation : public Operator {
 
   void close() override;
 
+  const std::vector<IdentityProjection>& identityProjections() const override {
```
Comment (mbasmanova) [inline velox/exec/HashAggregation.h:53]: Please, provide an analysis of why this is correct.

File: velox/exec/Operator.h (hunk 3 above)
Comment (xiaodouchen) [reply velox/exec/Operator.h]: From the code, identityProjections_ is only used `fillOutput()` and `identityProjections()`. In HashAggregation::getOutput(), fillOutput() is called only when isDistinct_ is true. Therefore, for the initialization of identityProjections_ within HashAgg, we can remove the restriction of isDistinct_ as it does not seem to have any side effects. What do you think?

File: velox/exec/Operator.h (hunk 5 above)
Comment (xiaodouchen) [reply velox/exec/Operator.h]: Has removed.

File: velox/exec/HashAggregation.h (hunk 6 above)
Comment (xiaodouchen) [reply velox/exec/HashAggregation.h]: It has been removed.

File: velox/exec/Operator.h (hunk 3 above)
Comment (mbasmanova) [reply velox/exec/Operator.h]: This analysis doesn't seem complete because identityProjections() is also used in Driver::pushdownFilters, hence, marking group-by keys as identity projections allows the filter pushdown to go through.

The change appears safe, but let's add some documentation for future readers.

- Operator.h - Let's add examples of identity projections, e.g. all columns in Filter operator, group-by keys in aggregation. Let's also clarify that these are used to enable filter pushdown.

File: velox/exec/tests/HashJoinTest.cpp [hunk 7]
```diff
@@ -5177,6 +5177,109 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+      }
+      SplitInput splits;
+      splits.emplace(nodeId, probeSplits);
+      return splits;
+    };
+  };
+
+  // Create build data
+  std::vector<RowVectorPtr> buildVectors;
+  for (int i = 0; i < 10; ++i) {
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5214]: Do we need 10 vectors for build side? I assume just a couple rows would be enough, no?

File: velox/exec/tests/HashJoinTest.cpp [hunk 8]
```diff
@@ -5177,6 +5177,109 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+        makeFlatVector<int64_t>(numRowsProbe, [](auto row) { return row; }),
+    });
+    probeVectors.push_back(rowVector);
+    tempFiles.push_back(TempFilePath::create());
+    writeToFile(tempFiles.back()->getPath(), rowVector);
+  }
+  auto makeInputSplits = [&](const core::PlanNodeId& nodeId) {
+    return [&] {
+      std::vector<exec::Split> probeSplits;
+      for (auto& file : tempFiles) {
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5202]: can we extract a helper function to make this test a bit shorter and easier to read?

```
return {{nodeId, makeSplits(tempFiles)}};
```

> exec::Split(makeHiveConnectorSplit(tempFiles.back()->getPath())));

Is it a bug to use 'tempFiles.back()' here?

File: velox/exec/tests/HashJoinTest.cpp [hunk 9]
```diff
@@ -5177,6 +5177,109 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
       .run();
 }
 
+TEST_F(HashJoinTest, dynamicFiltersPushDownThroughAgg) {
+  const int32_t numSplits = 10;
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5181]: Do we need to use that many splits?

File: velox/exec/tests/HashJoinTest.cpp [hunk 10]
```diff
@@ -5177,6 +5177,109 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+
+  // Create probe data
+  std::vector<RowVectorPtr> probeVectors;
+  probeVectors.reserve(numSplits);
+  std::vector<std::shared_ptr<TempFilePath>> tempFiles;
+  for (int32_t i = 0; i < numSplits; ++i) {
+    auto rowVector = makeRowVector({
+        makeFlatVector<int32_t>(
+            numRowsProbe, [&](auto row) { return row - i * 10; }),
+        makeFlatVector<int64_t>(numRowsProbe, [](auto row) { return row; }),
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5193]: nit: folly::identity

File: velox/exec/tests/HashJoinTest.cpp [hunk 11]
```diff
@@ -5177,6 +5177,109 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+  };
+
+  // Create build data
+  std::vector<RowVectorPtr> buildVectors;
+  for (int i = 0; i < 10; ++i) {
+    buildVectors.push_back(makeRowVector({
+        makeFlatVector<int32_t>(
+            numRowsBuild,
+            [i](auto row) { return 35 + 2 * (row + i * numRowsBuild / 5); }),
+        makeFlatVector<int64_t>(numRowsBuild, [](auto row) { return row; }),
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5219]: folly::identity

File: velox/exec/tests/HashJoinTest.cpp [hunk 12]
```diff
@@ -5177,6 +5177,109 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+  }
+
+  createDuckDbTable("t", probeVectors);
+  createDuckDbTable("u", buildVectors);
+
+  auto probeType = ROW({"c0", "c1"}, {INTEGER(), BIGINT()});
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  auto buildSide = PlanBuilder(planNodeIdGenerator, pool_.get())
+                       .values(buildVectors)
+                       .project({"c0 AS u_c0", "c1 AS u_c1"})
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5230]: You can remove this by using

```
buildVectors.push_back(makeRowVector({"u0", "u1"}, {..});
```

File: velox/exec/tests/HashJoinTest.cpp [hunk 13]
```diff
@@ -5177,6 +5177,109 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+                       .planNode();
+
+  // Inner join.
+  core::PlanNodeId probeScanId;
+  core::PlanNodeId joinId;
+  auto op = PlanBuilder(planNodeIdGenerator, pool_.get())
+                .tableScan(probeType)
+                .capturePlanNodeId(probeScanId)
+                .aggregation(
+                    {"c0", "c1"},
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5240]: .partialAggregation({"c0"}, {"sum(c1)"})

Is it a bug that c1 is a grouping key and used in sum(c1)?

File: velox/exec/tests/HashJoinTest.cpp [hunk 14]
```diff
@@ -5177,6 +5177,109 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+          ASSERT_EQ(0, getFiltersProduced(task, 2).sum);
+          ASSERT_EQ(0, getFiltersAccepted(task, 0).sum);
+          ASSERT_EQ(getInputPositions(task, 1), numRowsProbe * numSplits);
+          ASSERT_TRUE(planStats.at(probeScanId).dynamicFilterStats.empty());
+        } else {
+          ASSERT_EQ(2, getFiltersProduced(task, 2).sum);
+          ASSERT_EQ(2, getFiltersAccepted(task, 0).sum);
+          ASSERT_EQ(0, getReplacedWithFilterRows(task, 1).sum);
+          ASSERT_LT(getInputPositions(task, 1), numRowsProbe * numSplits);
+          ASSERT_EQ(
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5275]: > planStats.at(probeScanId).dynamicFilterStats

Can you extract a temp variable for this and re-use it in both branches for readability?

File: velox/exec/tests/HashJoinTest.cpp [hunk 15]
```diff
@@ -5177,6 +5177,109 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+              planStats.at(probeScanId).dynamicFilterStats.producerNodeIds,
+              std::unordered_set<core::PlanNodeId>({joinId}));
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5277]: > std::unordered_set<core::PlanNodeId>

Is this needed?

File: velox/exec/tests/HashJoinTest.cpp [hunk 16]
```diff
@@ -5177,6 +5177,109 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+        auto planStats = toPlanStats(task->taskStats());
+        if (hasSpill) {
+          // Dynamic filtering should be disabled with spilling triggered.
+          ASSERT_EQ(0, getFiltersProduced(task, 2).sum);
+          ASSERT_EQ(0, getFiltersAccepted(task, 0).sum);
+          ASSERT_EQ(getInputPositions(task, 1), numRowsProbe * numSplits);
+          ASSERT_TRUE(planStats.at(probeScanId).dynamicFilterStats.empty());
+        } else {
+          ASSERT_EQ(2, getFiltersProduced(task, 2).sum);
+          ASSERT_EQ(2, getFiltersAccepted(task, 0).sum);
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5272]: Why are we seeing that filters are not accepted? What's the point of pushing them down if they are ignored?

**Review by mbasmanova — CHANGES_REQUESTED** [review]

File: velox/exec/Operator.h (hunk 3 above)
Comment (xiaodouchen) [reply velox/exec/Operator.h]: Add comment of `identityProjections()`
```
  /// Returns a list of identity projections, e.g. columns that are projected
  /// as-is possibly after applying a filter. It is used to allow dynamic
  /// filters pushdown to go through. More examples of identity projections: all
  /// columns in FilterProject(only filters), group-by keys in HashAggregation.
```

File: velox/exec/tests/HashJoinTest.cpp (hunk 7 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp:5214]: Yes, don't need.

File: velox/exec/Operator.h [hunk 17]
```diff
@@ -454,7 +454,9 @@ class Operator : public BaseRuntimeStatWriter {
   }
 
   /// Returns a list of identity projections, e.g. columns that are projected
-  /// as-is possibly after applying a filter.
+  /// as-is possibly after applying a filter. It is used to allow dynamic
+  /// filters pushdown to go through. More examples of identity projections: all
```
Comment (mbasmanova) [inline velox/exec/Operator.h:458]: nit: "More examples" ->"Examples" since this is the first mentioning of examples.

File: velox/exec/Operator.h [hunk 18]
```diff
@@ -454,7 +454,9 @@ class Operator : public BaseRuntimeStatWriter {
   }
 
   /// Returns a list of identity projections, e.g. columns that are projected
-  /// as-is possibly after applying a filter.
+  /// as-is possibly after applying a filter. It is used to allow dynamic
```
Comment (mbasmanova) [inline velox/exec/Operator.h:457]: "It is used ..." -> "Used to allow pushdown of dynamic filters generated by HashProbe into the TableScan."

File: velox/exec/tests/HashJoinTest.cpp (hunk 8 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp]: `Is it a bug to use 'tempFiles.back()' here?` => Not a bug.
"makeInputSplits" has been removed.

File: velox/exec/tests/HashJoinTest.cpp (hunk 9 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp]: Not need

File: velox/exec/tests/HashJoinTest.cpp (hunk 10 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp]: Done.

File: velox/exec/tests/HashJoinTest.cpp (hunk 11 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp]: Done.

File: velox/exec/tests/HashJoinTest.cpp (hunk 12 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp]: Done.

File: velox/exec/tests/HashJoinTest.cpp (hunk 13 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp:5222]: Not a bug. It's valid in SQL. For example in test => "SELECT c0, c1, sum(c1) FROM t, u WHERE c0 = u0 and c1 = u1 group by c0, c1"

File: velox/exec/tests/HashJoinTest.cpp (hunk 14 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp]: Done.

File: velox/exec/tests/HashJoinTest.cpp (hunk 15 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp]: Yes. "std::unordered_set<core::PlanNodeId> producerNodeIds;" is defined in DynamicFilterStats.

File: velox/exec/tests/HashJoinTest.cpp (hunk 16 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp]: "0" in `getFiltersAccepted(task, 0)` is the OperatorIndex. Actually, it accepted 2 filters in table scan. I add a short comment here to make it more clear.

File: velox/exec/Operator.h (hunk 17 above)
Comment (xiaodouchen) [reply velox/exec/Operator.h]: Done.

File: velox/exec/Operator.h (hunk 18 above)
Comment (xiaodouchen) [reply velox/exec/Operator.h]: Done.

File: velox/exec/tests/HashJoinTest.cpp [hunk 19]
```diff
@@ -5177,6 +5177,80 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+  auto buildSide =
+      PlanBuilder(planNodeIdGenerator).values(buildVectors).planNode();
+
+  // Inner join.
+  core::PlanNodeId probeScanId;
+  core::PlanNodeId joinId;
+  auto op = PlanBuilder(planNodeIdGenerator, pool_.get())
+                .tableScan(probeType)
+                .capturePlanNodeId(probeScanId)
+                .partialAggregation({"c0", "c1"}, {"sum(c1)"})
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5221]: Why not invalid query, using c1 both as grouping key and as input to aggregation is confusing. Is there any particular reason you are constructing a plan like this? If so, would you add a comment to explain?

File: velox/exec/tests/HashJoinTest.cpp [hunk 20]
```diff
@@ -5177,6 +5177,80 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+      {probeScanId, {Split(makeHiveConnectorSplit(probeFile->getPath()))}}};
+  HashJoinBuilder(*pool_, duckDbQueryRunner_, driverExecutor_.get())
+      .planNode(std::move(op))
+      .inputSplits(splitInput)
+      .injectSpill(false)
+      .checkSpillStats(false)
+      .referenceQuery(
+          "SELECT c0, c1, sum(c1) FROM t, u WHERE c0 = u0 and c1 = u1 group by c0, c1")
+      .verifier([&](const std::shared_ptr<Task>& task, bool hasSpill) {
+        SCOPED_TRACE(fmt::format("hasSpill:{}", hasSpill));
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5242]: Do we need this? Spilling is always off.

File: velox/exec/tests/HashJoinTest.cpp [hunk 21]
```diff
@@ -5177,6 +5177,80 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+        auto planStats = toPlanStats(task->taskStats());
+        auto dynamicFilterStats = planStats.at(probeScanId).dynamicFilterStats;
+        ASSERT_EQ(2, getFiltersProduced(task, /* probeIndex */ 2).sum);
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5245]: It might be more readable to use plan node IDs instead of hard-coded indices.

File: velox/exec/tests/HashJoinTest.cpp [hunk 22]
```diff
@@ -5177,6 +5177,80 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+        makeFlatVector<int32_t>(
+            numRowsProbe, [&](auto row) { return row - 10; }),
+        makeFlatVector<int64_t>(numRowsProbe, folly::identity),
+    });
+  });
+  std::shared_ptr<TempFilePath> probeFile = TempFilePath::create();
+  writeToFile(probeFile->getPath(), probeVectors);
+
+  // Create build data
+  auto buildVectors = makeBatches(1, [&](int32_t /*unused*/) {
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5196]: No need to use makeBatches here, right?

File: velox/exec/tests/HashJoinTest.cpp [hunk 23]
```diff
@@ -5177,6 +5177,80 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
       .run();
 }
 
+TEST_F(HashJoinTest, dynamicFiltersPushDownThroughAgg) {
+  const int32_t numRowsProbe = 300;
+  const int32_t numRowsBuild = 100;
+
+  // Create probe data
+  auto probeVectors = makeBatches(1, [&](int32_t /*unused*/) {
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5185]: No need for makeBatches here, right?

**Review by mbasmanova — COMMENTED** [review]
@xiaodouchen Thank you for iterating. Some further comments about the test.

File: velox/exec/tests/HashJoinTest.cpp (hunk 23 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp:5185]: Yes

File: velox/exec/tests/HashJoinTest.cpp (hunk 22 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp:5196]: Yes

File: velox/exec/tests/HashJoinTest.cpp (hunk 20 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp:5242]: Has removed

File: velox/exec/tests/HashJoinTest.cpp (hunk 19 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp:5221]: Has removed it

File: velox/exec/tests/HashJoinTest.cpp (hunk 21 above)
Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp:5245]: Using plan node IDs as inputs may not be precise, as plan node IDs do not necessarily correspond one-to-one with operator indices.

Comment (xiaodouchen) [reply velox/exec/tests/HashJoinTest.cpp]: Thanks for the detailed explanation. I have added `getOperatorIndex()` to help get the operator index from the plan node id.
```
  // Get the operator index from the plan node id. Only used in the probe-side
  // pipeline. The plan node id starts from "1" and the operator index starts
  // from 0. Plan node IDs map to operators 1:1.
  static int32_t getOperatorIndex(const core::PlanNodeId& planNodeId) {
    return folly::to<int32_t>(planNodeId) - 1;
  }
```

File: velox/exec/tests/HashJoinTest.cpp [hunk 24]
```diff
@@ -5177,6 +5184,75 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
...
+      .planNode(std::move(op))
+      .inputSplits(splitInput)
+      .injectSpill(false)
+      .checkSpillStats(false)
+      .referenceQuery("SELECT c0, sum(c1) FROM t, u WHERE c0 = u0 group by c0")
+      .verifier([&](const std::shared_ptr<Task>& task, bool hasSpill) {
+        auto planStats = toPlanStats(task->taskStats());
+        auto dynamicFilterStats = planStats.at(scanNodeId).dynamicFilterStats;
+        ASSERT_EQ(
+            1, getFiltersProduced(task, getOperatorIndex(joinNodeId)).sum);
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:5244]: I feel getFiltersProduced(joinNodeId) is more readable. CC: @xiaoxmeng

**Review by mbasmanova — APPROVED** [review]
